pkg/util/ip_reservation_test.go

```diff
@@ -0,0 +1,183 @@
+	// Reserved IPV4 ranges listed by the cloud provider
+	reservedIPV4Ranges := make(map[string]bool)
+	cases := []struct {
+		name          string
+		expected      string
+		errorExpected bool
+	}{
+		{
+			name:          "0/4 /29 IP blocks used",
+			expected:      "192.168.92.0/29",
+			errorExpected: false,
+		}, {
+			name:          "1/4 /29 IP blocks used",
+			expected:      "192.168.92.8/29",
+			errorExpected: false,
+		},
+		{
+			name:          "2/4 /29 IP blocks used",
+			expected:      "192.168.92.16/29",
+			errorExpected: false,
+		},
+		{
+			name:          "3/4 /29 IP blocks used",
+			expected:      "192.168.92.24/29",
+			errorExpected: false,
+		},
+		{
+			name:          "All /29 IP blocks used",
+			expected:      "",
+			errorExpected: true,
+		},
+	}
+
+	for index, test := range cases {
+		ipAllocator := initTestIPAllocator()
+		for i := 0; i < index; i++ {
+			// Distributing IPs between the reserved lists at the cloud provider end and the reserved list in the ipAllocator
+			if i%2 == 0 {
+				ipAllocator.reserved[ips[i]] = true
+			} else {
+				reservedIPV4Ranges[ips[i]] = true
+			}
+
+		}
+
+		ipBlock, err := ipAllocator.GetUnreservedIPBlock(cidr, reservedIPV4Ranges)
+		if err != nil && !test.errorExpected {
+			t.Errorf("test %q failed: got error %s, expected %s", test.name, err.Error(), test.expected)
+		} else if err == nil && test.errorExpected {
+			t.Errorf("test %q failed: got reserved IP range %s, expected error", test.name, ipBlock)
+		} else if ipBlock != test.expected {
+			t.Errorf("test %q failed: got reserved IP range %s, expected %s", test.name, ipBlock, test.expected)
+		}
+	}
+}
+func TestValidateCIDROverlap(t *testing.T) {
+	cases := []struct {
+		name     string
+		cidr1    string
+		cidr2    string
+		expected bool
+	}{
+		{
+			name:     "Overlapping CIDRs",
+			cidr1:    "192.168.92.0/29",
+			cidr2:    "192.168.92.48/26",
+			expected: true,
+		},
+		{
+			name:     "Non overlapping CIDRs",
+			cidr1:    "192.168.92.0/29",
+			cidr2:    "192.168.22.67/26",
+			expected: false,
+		},
+	}
+
+	for _, test := range cases {
+		overlap := isOverlap(test.cidr1, test.cidr2)
+		if overlap != test.expected {
+			t.Errorf("test %q failed: got CIDR overlap  as %t, expected %t", test.name, overlap, test.expected)
+		}
+	}
+}
+
+func TestIncrementIP(t *testing.T) {
+	step := byte(8)
+
+	cases := []struct {
+		name          string
+		currentIP     string
+		expected      string
+		errorExpected bool
+	}{
+		{
+			name:          "Valid IP increment without any carry forward to significant bytes",
+			currentIP:     "192.168.92.32",
+			expected:      "192.168.92.40",
+			errorExpected: false,
+		}, {
+			name:          "Valid increment with carry forward to significant bytes",
+			currentIP:     "192.168.255.253",
+			expected:      "192.169.0.5",
+			errorExpected: false,
+		},
+		{
+			name:          "Invalid increment",
+			currentIP:     "255.255.255.253",
+			expected:      "0.0.0.5",
```

Comment:
nit: Don't specify an expected answer if there is none (we expect an error here)